### PR TITLE
create a file to hold documentation of import-boss in one place

### LINKED ISSUE #81804: Document the use of import-boss
[import-boss](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/code-generator/cmd/import-boss/main.go) is used in our CI tests ([./hack/verify-import-boss.sh](https://github.com/kubernetes/kubernetes/blob/master/hack/verify-import-boss.sh)) to verify import restrictions in various places throughout k/k.
Currently, the only existing documentation is within the script itself and this.

xref: https://github.com/kubernetes/kubernetes/issues/81102 - import-boss ignores import in a _test.go file
xref: https://github.com/kubernetes/kubernetes/issues/81803 - failed to implement verify import restrictions in the e2 test framework.

Seems that import-boss is owned by API machinery
/sig api-machinery
/area code-organization
/kind documentation
/priority important-longterm
/cc @jaypipes

### PR DESCRIPTION
**Pull request by MorrisLaw**
**What type of PR is this?**
/kind documentation

**What this PR does / why we need it**:
It's been noted that documentation for `import-boss` is lacking/spread out. 
- Moved the current documentation from a comment in the `import-boss/main.go` file to a separate markdown file within the `import-boss` directory. 
- Added additional information on how to run it with and without the flag to include `_test.go` files.
- Added additional information regarding known "gotchas" that have been talked about in various issue comments/PR comments.

**Which issue(s) this PR fixes**:

Ref #81804

**Special notes for your reviewer**:
I am working with @alejandrox1 on what needs to be documented. Made this PR to track our progress.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/code-generator/cmd/import-boss/README.md
Comment (neolit123) [inline staging/src/k8s.io/code-generator/cmd/import-boss/README.md]: seems better to also quote ".import-restrictions" like so `.import-restrictions`

Comment (neolit123) [reply staging/src/k8s.io/code-generator/cmd/import-boss/README.md]: isn't --input-dirs also a flag on the command line?

File: staging/src/k8s.io/code-generator/cmd/import-boss/README.md
Comment (neolit123) [inline staging/src/k8s.io/code-generator/cmd/import-boss/README.md]: quote `SelectorRegexp`, `AllowedPrefixes`, `ForbiddenPrefixes`

File: staging/src/k8s.io/code-generator/cmd/import-boss/README.md
Comment (neolit123) [inline staging/src/k8s.io/code-generator/cmd/import-boss/README.md]: "any forbidden prefix" -> "any forbidden prefixes"

File: staging/src/k8s.io/code-generator/cmd/import-boss/README.md
Comment (MorrisLaw) [reply staging/src/k8s.io/code-generator/cmd/import-boss/README.md]: Yes it is. Nice catch 👍

File: staging/src/k8s.io/code-generator/cmd/import-boss/README.md
Comment (alejandrox1) [inline staging/src/k8s.io/code-generator/cmd/import-boss/README.md]: being super verbose but could you specify that `arguments.InoutDirs` is defined in the src code (main.go file in the same dir)

File: staging/src/k8s.io/code-generator/cmd/import-boss/README.md
Comment (alejandrox1) [inline staging/src/k8s.io/code-generator/cmd/import-boss/README.md]: nit:
```suggestion
- An import is allowed if it matches at least one allowed prefix and does not match any forbidden prefixes. An example `.import-restrictions` file looks like this:
```

**Review by alejandrox1 — COMMENTED** [review]
Thank you for working on this @MorrisLaw !
/priority backlog

**Comment by MorrisLaw** [comment]
Hey @alejandrox1  @neolit123 can I please get another review for the recent changes made to address the comments. Thank you!

**Review by alejandrox1 — APPROVED** [review]
This looks good!
Thank you
/lgtm

/assign @wojtek-t 
could you take a look please?

**Comment by MorrisLaw** [comment]
Hey @wojtek-t, friendly ping to review this doc PR

File: staging/src/k8s.io/code-generator/cmd/import-boss/README.md
Comment (neolit123) [inline staging/src/k8s.io/code-generator/cmd/import-boss/README.md]: i'd indent the second line the same as the previous one ` - When....`
and remove the `NOTE: ` prefix:

```
- When a directory is verified...
- Adding `.import-restrictions`...
```

File: staging/src/k8s.io/code-generator/cmd/import-boss/README.md
Comment (neolit123) [inline staging/src/k8s.io/code-generator/cmd/import-boss/README.md]: possibly word this, like so?
> If this file is not found, this directory and it's sub-directories will be skipped.

File: staging/src/k8s.io/code-generator/cmd/import-boss/README.md
Comment (neolit123) [inline staging/src/k8s.io/code-generator/cmd/import-boss/README.md]: `the unsafe package` -> `the "unsafe" package`

i'd also explain with a sentence how `"SelectorRegexp": "k8s[.]io"` is handled above.

Comment (neolit123) [reply staging/src/k8s.io/code-generator/cmd/import-boss/README.md]: possibly omit the `NOTE:` prefix again.

**Comment by neolit123** [comment]
thanks for working on this document @MorrisLaw 
i have added a couple of minor comments.

**Comment by MorrisLaw** [comment]
> thanks for working on this document @MorrisLaw
> i have added a couple of minor comments.

Np and thanks for the review! Just pushed up some changes @neolit123 .

File: staging/src/k8s.io/code-generator/cmd/import-boss/README.md
Comment (neolit123) [inline staging/src/k8s.io/code-generator/cmd/import-boss/README.md]: possibly use back-tic to quote `"SelectorRegexp": "k8s[.]io"`

File: staging/src/k8s.io/code-generator/cmd/import-boss/README.md
Comment (neolit123) [inline staging/src/k8s.io/code-generator/cmd/import-boss/README.md]: `it's` -> `its`

this is a type that came from my suggestion, sorry.
https://github.com/kubernetes/kubernetes/pull/85564#discussion_r392239151

File: staging/src/k8s.io/code-generator/cmd/import-boss/README.md
Comment (neolit123) [inline staging/src/k8s.io/code-generator/cmd/import-boss/README.md]: > In order for them to be added, they need to be explicitly added to `arguments.InputDirs`.

->

> They need to be explicitly added to `arguments.InputDirs`.

**Comment by neolit123** [comment]
better to keep the commits squashed to 1, using rebase and `git push -f`

**Comment by alejandrox1** [comment]
More git tricks https://blog.sebastian-daschner.com/entries/git-commit-fixup-autosquash

File: staging/src/k8s.io/code-generator/cmd/import-boss/README.md
Comment (oomichi) [reply staging/src/k8s.io/code-generator/cmd/import-boss/README.md]: `this directory and it's sub-directories will be skipped.`
If sub-directories contains .import-restrictions file, they are checked with the file.
So we can say just `this directory will be skipped.`

File: staging/src/k8s.io/code-generator/cmd/import-boss/README.md
Comment (oomichi) [reply staging/src/k8s.io/code-generator/cmd/import-boss/README.md]: +1

**Review by oomichi — COMMENTED** [review]
Thanks for doing this, really nice work.
I wanted to read this before fixing import-boss for e2e framework.
just few comments on that.

File: staging/src/k8s.io/code-generator/cmd/import-boss/README.md
Comment (oomichi) [reply staging/src/k8s.io/code-generator/cmd/import-boss/README.md]: After the commit 558f6aab56c0e983741d7df6eb8ffc79a19a8672 we need to specify directories which contain `.import-restrictions` on hack/verify-import-boss.sh instead of import-boss/main.go.
So we need to update here.

/cc @oomichi

Comment (MorrisLaw) [reply staging/src/k8s.io/code-generator/cmd/import-boss/README.md]: @neolit123 I'm interpreting this comment as, replace the former with the latter?

Comment (neolit123) [reply staging/src/k8s.io/code-generator/cmd/import-boss/README.md]: my comment is no longer valid. based on @oomichi 's comment i think we should formulate this in the lines of:

> - To add more paths to be checked by the k8s CI, add them in the file `hack/verify-import-boss.sh

its here:
https://github.com/kubernetes/kubernetes/blob/master/hack/verify-import-boss.sh#L33

**Comment by MorrisLaw** [comment]
Thanks for the thorough reviewing @neolit123 @oomichi , just pushed up new changes 👍

File: staging/src/k8s.io/code-generator/cmd/import-boss/README.md
Comment (neolit123) [inline staging/src/k8s.io/code-generator/cmd/import-boss/README.md]: i'd remove this example and leave it to the users to look at the file and find the packages variable.
it might change in the future so better not document it explicitly.

File: staging/src/k8s.io/code-generator/cmd/import-boss/README.md
Comment (neolit123) [inline staging/src/k8s.io/code-generator/cmd/import-boss/README.md]: the main.go examples should be removed now?

File: staging/src/k8s.io/code-generator/cmd/import-boss/README.md
Comment (MorrisLaw) [reply staging/src/k8s.io/code-generator/cmd/import-boss/README.md]: Good point. Makes the document more resilient. Will update.

**Comment by neolit123** [comment]
thanks for the updates @MorrisLaw 
looks like approvers are already assigned.

/lgtm
cc @sttts @wojtek-t

**Comment by oomichi** [comment]
Thanks for updating @MorrisLaw 
This is helpful to understand who to maintain import-boss check.

/lgtm

File: staging/src/k8s.io/code-generator/cmd/import-boss/README.md
Comment (sttts) [inline staging/src/k8s.io/code-generator/cmd/import-boss/README.md]: this section is a little odd here because it only applies to the k/k repo. At least mention that.

background: this file will show up at https://github.com/kubernetes/code-generator/tree/master/cmd/import-boss too.

Comment (MorrisLaw) [reply staging/src/k8s.io/code-generator/cmd/import-boss/README.md]: Something like this @sttts ?

`## How do I run import-boss within the k/k repo?`

Comment (sttts) [reply staging/src/k8s.io/code-generator/cmd/import-boss/README.md]: Yes.

**Comment by MorrisLaw** [comment]
@sttts final review/thumbs for this PR please

File: staging/src/k8s.io/code-generator/cmd/import-boss/README.md
Comment (sttts) [inline staging/src/k8s.io/code-generator/cmd/import-boss/README.md]: actually it goes up to the parent dir until it finds one, @eloyekunle can comment

File: staging/src/k8s.io/code-generator/cmd/import-boss/README.md
Comment (sttts) [inline staging/src/k8s.io/code-generator/cmd/import-boss/README.md]: is this true? cc @eloyekunle

File: staging/src/k8s.io/code-generator/cmd/import-boss/README.md
Comment (eloyekunle) [reply staging/src/k8s.io/code-generator/cmd/import-boss/README.md]: Yes, it goes up to the parent directory.

File: staging/src/k8s.io/code-generator/cmd/import-boss/README.md
Comment (eloyekunle) [reply staging/src/k8s.io/code-generator/cmd/import-boss/README.md]: Once a parent directory has been added to `hack/verify-import-boss.sh`, the subpackages do not need to be added but will be checked automatically.

Currently, everything in `staging/src/k8s.io/*/`, as well as `k8s.io/kubernetes/pkg/..., k8s.io/kubernetes/cmd/..., k8s.io/kubernetes/plugin/..., k8s.io/kubernetes/test/e2e/framework/...` are currently checked, and that includes most of the code in k/k.

**Comment by MorrisLaw** [comment]
Based on comments by @sttts and @eloyekunle and @eloyekunle's [PR](https://github.com/kubernetes/community/pull/4529). I've added more updates to this doc. Please review again.

**Comment by eloyekunle** [comment]
/lgtm

cc @sttts

**Comment by MorrisLaw** [comment]
@sttts can we merge this in?

**Comment by MorrisLaw** [comment]
@wojtek-t @lavalamp ^

File: staging/src/k8s.io/code-generator/cmd/import-boss/main.go
Comment (lavalamp) [inline staging/src/k8s.io/code-generator/cmd/import-boss/main.go:17]: I'd either leave this, or at least keep the top line (so that something shows up in godoc).

File: staging/src/k8s.io/code-generator/cmd/import-boss/README.md
Comment (lavalamp) [inline staging/src/k8s.io/code-generator/cmd/import-boss/README.md:53]: Can you expand? Why are inverse rules useful? When would one use the transitive option?

(thanks for fixing the documentation!)

**Comment by MorrisLaw** [comment]
Thanks for reviewing @lavalamp , I've made some changes to address your comments. Wdyt of the updates?

**Comment by MorrisLaw** [comment]
@oomichi @eloyekunle @sttts @neolit123 made some updates based on recent comments ^

**Pull request merged**
